src/Nethermind/Nethermind.TxPool/Collections/PersistentBlobTxDistinctSortedPool.cs

```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+
+namespace Nethermind.TxPool.Collections;
+
+public class PersistentBlobTxDistinctSortedPool : BlobTxDistinctSortedPool
+{
+    private readonly ITxStorage _blobTxStorage;
+    private readonly LruCache<ValueKeccak, Transaction> _blobTxCache;
+    private readonly ILogger _logger;
+
+    public PersistentBlobTxDistinctSortedPool(ITxStorage blobTxStorage, ITxPoolConfig txPoolConfig, IComparer<Transaction> comparer, ILogManager logManager)
+        : base(txPoolConfig.PersistentBlobStorageSize, comparer, logManager)
+    {
+        _blobTxStorage = blobTxStorage ?? throw new ArgumentNullException(nameof(blobTxStorage));
+        _blobTxCache = new(txPoolConfig.BlobCacheSize, txPoolConfig.BlobCacheSize, "blob txs cache");
+        _logger = logManager.GetClassLogger();
+
+        RecreateLightTxCollectionAndCache(blobTxStorage);
+    }
+
+    private void RecreateLightTxCollectionAndCache(ITxStorage blobTxStorage)
+    {
+        if (_logger.IsDebug) _logger.Debug("Recreating light collection of blob transactions and cache");
+        foreach (Transaction fullBlobTx in blobTxStorage.GetAll())
```

Comment:
If we want to have proper sorting, we don't have a choice. Light collection is the first place where we are checking if we have tx - otherwise we would need to always call db, which is not a good idea. Also it is used for sorting - we will not know which tx should be evicted if pool size will be exceeded, another thing that we will not know which transactions are the best when building blocks. Recreation will be done once, on starting the node and with current design we just need this collection. I did loading from db as lazy enumeration, so we shouldn't have memory spike, we will just get tx from db, save what we need in memory and move to another tx

Reply:
It's refactored to read from db only column of light blob txs, now it is recreating on restart in max 100-150ms instead of 30-40s in case of full blob pool of 6-blob txs so worst case scenario